wolfcrypt/src/port/Renesas/renesas_tsip_aes.c

```diff
@@ -151,6 +168,349 @@ int wc_tsip_AesCbcDecrypt(struct Aes* aes, byte* out, const byte* in, word32 sz)
+    }
+
+    userCtx = (TsipUserCtx *)ctx;
 
+    /* buffer for plain data outout must be multiple of AES_BLOCK_SIZE */
+    plainBufSz = ((sz / AES_BLOCK_SIZE) + 1) * AES_BLOCK_SIZE;
+
+    if ((ret = tsip_hw_lock()) == 0) {
+
+        /* allocate buffers for plain-taxt, cipher-text, authTag and AAD.
+         * TSIP requests those buffers 32bit aligned. 
+         */
+        cipherBuf = XMALLOC(sz, aes->heap, DYNAMIC_TYPE_AES);
+        plainBuf  = XMALLOC(plainBufSz, aes->heap, DYNAMIC_TYPE_AES);
+        aTagBuf   = XMALLOC(TSIP_AES_GCM_AUTH_TAG_SIZE, aes->heap, 
+                                                        DYNAMIC_TYPE_AES);
+        aadBuf    = XMALLOC(authInSz, aes->heap, DYNAMIC_TYPE_AES);
+
+        if (plainBuf == NULL || cipherBuf == NULL || aTagBuf == NULL ||
+                                                        aadBuf == NULL) {
+            ret = -1;
+        }
+
+        if (ret == 0) {
+            XMEMSET(plainBuf,  0, plainBufSz);
+            XMEMCPY(cipherBuf, in, sz);
+            XMEMSET(aTagBuf, 0, TSIP_AES_GCM_AUTH_TAG_SIZE);
+            XMEMCPY(aTagBuf,authTag,min(authTagSz, TSIP_AES_GCM_AUTH_TAG_SIZE));
+            XMEMCPY(aadBuf, authIn, authInSz);
+        }
+
+        if (ret == 0) {
+            /* generate AES-GCM session key. The key stored in 
+             * Aes.ctx.tsip_keyIdx is not used here. 
+             */
+            err = R_TSIP_TlsGenerateSessionKey(
+                    userCtx->tsip_cipher,
+                    (uint32_t*)userCtx->tsip_masterSecret,
+                    (uint8_t*) userCtx->tsip_clientRandom,
+                    (uint8_t*) userCtx->tsip_serverRandom,
+                    (uint8_t*)&iv[AESGCM_IMP_IV_SZ], /* use exp_IV */
+                    NULL,
+                    NULL,
+                    NULL,
+                    &key_server_aes,
+                    NULL, NULL);
+            if (err != TSIP_SUCCESS) {
+                WOLFSSL_MSG("R_TSIP_TlsGenerateSessionKey failed");
+                ret = -1;
+            }
+        }
+
+        if (ret == 0) {
+            /* since key_index has iv and ivSz in it, no need to pass them init 
+             * func. Pass NULL and 0 as 3rd and 4th parameter respectively.
+             */
+            err = initFn(&hdl, &key_server_aes, NULL, 0UL);
+
+            if (err == TSIP_SUCCESS) {
+                /* pass only AAD and it's size before passing cipher text */
+                err = updateFn(&hdl, NULL, NULL, 0UL, (uint8_t*)authIn, 
+                                                                    authInSz);
+            }
+            if (err == TSIP_SUCCESS) {
+                err = updateFn(&hdl, cipherBuf, plainBuf, sz, NULL, 0UL);
+            }
+            if (err != TSIP_SUCCESS) {
+                WOLFSSL_MSG("R_TSIP_AesXXXGcmDecryptUpdate: failed in decrypt");
+                ret = -1;
+            }
+
+            dataLen = 0;
+            err = finalFn(&hdl,
```

Comment:
Should we check `if (err == TSIP_SUCCESS)` before calling finalFn?

Reply:
Please address

Reply:
Added check.